libraries/ms-common/src/stm32f0xx/pwm_input.c

```diff
@@ -0,0 +1,144 @@
+};
+
+StatusCode pwm_input_init(PwmTimer timer, PwmChannel channel) {
+  s_port[timer].rcc_cmd(s_port[timer].periph, ENABLE);
+
+  uint16_t trigger_source = 0;
+
+  // Sets the trigger source depending on whether we're using channel 1 or 2
+  if (channel == PWM_CHANNEL_1) {
+    s_port[timer].channel = TIM_Channel_1;
+    trigger_source = TIM_TS_TI1FP1;
+  } else {
+    s_port[timer].channel = TIM_Channel_2;
+    trigger_source = TIM_TS_TI2FP2;
+  }
+
+  TIM_TypeDef *tim_location = s_port[timer].base;
+
+  // Gets our current frequency
+  RCC_ClocksTypeDef clocks;
+  RCC_GetClocksFreq(&clocks);
+
+  // Struct to configure PWM frequency (this will determine what unit the user gets)
+  TIM_TimeBaseInitTypeDef tim_init = {
+    .TIM_Prescaler = (clocks.PCLK_Frequency / 1000000) - 1,
+    .TIM_CounterMode = TIM_CounterMode_Up,
+    .TIM_Period = 0xFFFFFFFF,
+    .TIM_ClockDivision = TIM_CKD_DIV1,
+    .TIM_RepetitionCounter = 0,
+  };
+
+  TIM_TimeBaseInit(tim_location, &tim_init);
+
+  // Struct to configure the timer for PWM input mode
+  TIM_ICInitTypeDef tim_icinit = {
+    .TIM_Channel = s_port[timer].channel,
+    .TIM_ICPolarity = TIM_ICPolarity_Rising,
+    .TIM_ICSelection = TIM_ICSelection_DirectTI,
+    .TIM_ICPrescaler = TIM_ICPSC_DIV1,
+    .TIM_ICFilter = 0x0,
+  };
+
+  TIM_PWMIConfig(tim_location, &tim_icinit);
+
+  // Puts the timer into PWM input mode
+  TIM_SelectInputTrigger(tim_location, trigger_source);
+  TIM_SelectSlaveMode(tim_location, TIM_SlaveMode_Reset);
+  TIM_SelectMasterSlaveMode(tim_location, TIM_MasterSlaveMode_Enable);
+  TIM_SelectOutputTrigger(tim_location, TIM_TRGOSource_Reset);
+  TIM_Cmd(tim_location, ENABLE);
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode pwm_input_get_reading(PwmTimer timer, PwmInputReading *reading) {
+  if (reading == NULL) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "|reading cannot be null|\n");
+  }
+
+  TIM_TypeDef *tim_location = s_port[timer].base;
+
+  // Returns reading if the Capture Compare flag is set, otherwise
+  // returns a PWM of 0
+  if (TIM_GetFlagStatus(tim_location, TIM_FLAG_CC1) == SET) {
+    TIM_ClearFlag(tim_location, TIM_FLAG_CC1);
+  } else {
+    reading->dc = 0;
+    reading->period_us = 0;
+    return STATUS_CODE_OK;
+  }
+
+  uint32_t IC2Value_1 = 0;
+  uint32_t IC2Value_2 = 0;
+
+  uint32_t period_us = 0;
+  uint32_t dc = 0;
+
+  // Depending on which channel we use, the values need to be flipped
+  if (s_port[timer].channel == TIM_Channel_2) {
+    IC2Value_1 = TIM_GetCapture2(tim_location);
+    IC2Value_2 = TIM_GetCapture1(tim_location);
+  } else {
+    IC2Value_2 = TIM_GetCapture2(tim_location);
+    IC2Value_1 = TIM_GetCapture1(tim_location);
+  }
+
+  // Perform the PWM calculation. IC2Value_1 is the period, and IC2Value_2 is the time
+  // that the signal is high
+  if (IC2Value_1 != 0) {
+    dc = (IC2Value_2 * 1000) / IC2Value_1;
+    period_us = IC2Value_1;
+  } else {
+    dc = 0;
+    period_us = 0;
+  }
+
+  if (dc > MAX_DC_VALUE) {
```

Comment:
Why are we doing this?

Reply:
DC is a percentage value between 0 and 100. I just did some sanitation so that we're not returning values greater than 100 (actually 1000 since we're returning the percentage value * 10).

Reply:
OK can we note that somewhere? Like in the `typdef` of the struct or change the name from `dc` to like `dc_percent`. If we're doing that also probably change to `MAX_DC_PERCENT = 1000`